_posts/2017-07-29-ruby-oop.md

```diff
@@ -0,0 +1,381 @@
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, as it is already understood what you mean.
+
+Our new `#distance` method will take another `Point` object as an argument, read its `@x` and `@y` attributes, and do the maths to return our result implicitly by being the last line of the method.
+
+Let's try it out:
+```ruby
+point_a = Point.new(1.5, -3.0)
+point_b = Point.new(2, 1.0)
+
+puts point_a.distance point_b #=> 4.031128874149275
+```
+
+**We can now:**
+
+1. Describe a `Point`.
+2. Measure the distance between any two `Point`s.
+
+### Line
+
+Let's refer back to our definition of a `Line`.
+
+```ruby
+# A finite segment (or distance) between two points.
+class Line
+end
+```
+
+From what we know, we can now give our `Line`s two attributes. We'll call them `start` and `stop` (I would use the word `end` here, but this is a Ruby keyword and it could get confusing.)
+
+```ruby
+# A finite segment (or distance) between two points.
+class Line
+  attr_reader :start
+
+  attr_reader :stop
+
+  def initialize(start, stop)
+    @start = start
+    @stop = stop
+  end
+end
+```
+
+Great. What other properties does a finite line have? Ah, right - length.
+
+But - we've already implemented this! We already know the distance between any two `Point` objects.. let's use it:
+
+```ruby
+# A finite segment (or distance) between two points.
+class Line
+  def length
+    start.distance stop
+  end
+end
+```
+I'll use the same two points from before and test it:
+```ruby
+point_a = Point.new(1.5, -3.0)
+point_b = Point.new(2, 1.0)
+
+line = Line.new(point_a, point_b)
+puts line.length #=> 4.031128874149275
+```
+> When talking about Ruby code, what we've done here is "wrapped" the `#distance` behavior of a `Point` object - or rather, we "abstracted" it - to give us the length of any kind of line we could describe.
+
+**Now we can:**
+
+1. Describe a `Point`.
+2. Measure the distance between any two `Point`s.
+3. Create a `Line` from any two `Point`s.
+4. Measure the `#length` of that `Line`.
+
+Take note of how we described the problem in simplest terms, and are slowly building up *layers* of objects and methods in order to craft our solution.
+
+This leaves one last piece to our puzzle.
+
+### Polygon
+
+We can now move on to describe a `Polygon` as a collection of `Line`s.
+
+```ruby
+# A collection of lines that describe a 2D shape.
+class Polygon
+  attr_reader :segments
+
+  def initialize(segments = [])
+    @segments = segments
+  end
+end
+```
+> Arguments for methods can be given a default value using the above syntax in `def initialize`. Here, I'm making `segments` default to an empty array.
+
+We're going to design `Polygon` to accept an `Array` of `Line`.
```

Comment:
I think that may be out of the scope of the tutorial (basic, high level OOP design) but I can add a small note. (i.e., this isn't a *beginner's Ruby* tutorial, it's an OOP tutorial)